### PR TITLE
FE-705 | Refactor parsing logic to be order agnostic

### PR DESCRIPTION
**Pull request by fauna-ryanharris**
### Notes
[Jira Ticket](https://faunadb.atlassian.net/browse/FE-705)

Currently, the JS driver depends on the order of the arguments it receives as wire protocol, which should not be the case. We should not base the logic on order, so this PR attempts to improve that functionality.

### How to test
Run `npm run test`

### CODE REVIEW

File: test/values.test.js [hunk 1]
```diff
@@ -150,6 +150,49 @@ describe('Values', () => {
...
+    var complexQuery = new Query({
+      lambda: '_',
+      expr: { match: { index: 'idx' }, terms: ['str', 10] },
+      api_version: '3',
+    })
+
+    // simpleQuery assertions
+    var x = '{"@query":{"api_version":"3","lambda":"x","expr":{"var":"x"}}}'
+    expect(json.parseJSON(x)).toEqual(simpleQuery)
+    expect(json.parseJSON(x).toString()).toEqual(simpleQuery.toString())
```
Comment (fauna-ryanharris) [inline test/values.test.js:169]: @marrony Would it be better to have these `.toString()` assertion in the `pretty print` test block below?

Comment (marrony) [reply test/values.test.js:169]: I think the `.toString()` assertions here are not necessary, if you parsed the json and it equals to what you expect, then the `.toString()` would be the same

Comment (fauna-ryanharris) [reply test/values.test.js:169]: @marrony That's what I was thinking as well -- just trying to cover each case as well as I can. I'll remove these 👍

File: src/Expr.js [hunk 2]
```diff
@@ -169,35 +236,99 @@ var exprToString = function(expr, caller) {
...
+    )
+  }
+
+  if ('call' in expr) {
+    if (Array.isArray(expr.arguments.raw)) {
+      return (
+        'Call(' +
+        exprToString(expr['call']) +
+        ', ' +
+        printArray(expr['arguments']['raw'], exprToString) +
```
Comment (marrony) [inline src/Expr.js:297]: instead of introspect `raw` field here, you should call the function `exprToString` recursively and it should resolve the value, in the beginning of the function it already does this check

File: test/values.test.js [hunk 3]
```diff
@@ -150,6 +150,86 @@ describe('Values', () => {
...
+    var b =
+      '{"@query":{"expr":{"match":{"index":"idx"},"terms":["str",10]},"api_version":"3","lambda":"_"}}'
+    expect(json.parseJSON(b)).toEqual(complexQuery)
+
+    var c =
+      '{"@query":{"api_version":"3","expr":{"match":{"index":"idx"},"terms":["str",10]},"lambda":"_"}}'
+    expect(json.parseJSON(c)).toEqual(complexQuery)
+  })
+
+  test('parse functions regardless of order', () => {
```
Comment (marrony) [inline test/values.test.js:190]: this test is not necessary because the driver doesn't parse the expression, you should focus on the pretty print tests to ensure the queries values render the correct code

Comment (fauna-ryanharris) [reply test/values.test.js:190]: @marrony I was going to ask you about that 😅 Should I just remove this whole `test` block?

Comment (marrony) [reply test/values.test.js:190]: fell free to remove

File: src/Expr.js [hunk 4]
```diff
@@ -169,35 +236,109 @@ var exprToString = function(expr, caller) {
...
+      ', ' +
+      exprToString(expr['else']) +
+      ')'
+    )
+  }
+
+  if ('call' in expr) {
+    var argumentsExpr = null
+
+    if (isExpr(expr['arguments'])) {
```
Comment (marrony) [inline src/Expr.js:294]: this whole logic is not necessary, just calling `argumentsExpr = exprToString(expr['arguments'])` should be enough, all the logic added here was already handled in the beginning of the function

File: src/Expr.js [hunk 5]
```diff
@@ -169,35 +236,109 @@ var exprToString = function(expr, caller) {
...
+        argumentsExpr = '[' + printArray(expr['arguments'], exprToString) + ']'
+      } else if (typeof expr['arguments'] === 'object') {
+        argumentsExpr = printObject(expr['arguments'])
+      } else {
+        argumentsExpr = exprToString(expr['arguments'])
+      }
+    }
+
+    var callExpr = ''
+    if (isExpr(expr['call'])) {
```
Comment (marrony) [inline src/Expr.js:307]: no need to check if `expr['call']` is an expression, the recursion will already do this

File: src/Expr.js [hunk 6]
```diff
@@ -169,35 +236,109 @@ var exprToString = function(expr, caller) {
...
+      callExpr = exprToString(expr['call'])
+    } else {
+      callExpr = exprToString(expr['call'])
+    }
+
+    return 'Call(' + callExpr + ', ' + argumentsExpr + ')'
```
Comment (marrony) [inline src/Expr.js:313]: this can be simplified by
```javascript
return 'Call(' + exprToString(expr['call']) + ', ' + exprToString(expr['arguments']) + ')'
```

File: src/Expr.js (hunk 5 above)
Comment (fauna-ryanharris) [reply src/Expr.js:299]: Ah, true! Thanks. Will update.

File: src/Expr.js (hunk 4 above)
Comment (fauna-ryanharris) [reply src/Expr.js:286]: I added this check because one of the pre-existing unit tests was getting to this point where the `arguments` and `call` values were both expressions 🤔

File: src/Expr.js [hunk 7]
```diff
@@ -169,35 +228,109 @@ var exprToString = function(expr, caller) {
...
+    if (isExpr(expr['arguments'])) {
+      argumentsExpr = exprToString(expr['arguments'])
+    } else {
+      if (Array.isArray(expr['arguments'])) {
+        argumentsExpr = '[' + printArray(expr['arguments'], exprToString) + ']'
+      } else if (typeof expr['arguments'] === 'object') {
+        argumentsExpr = printObject(expr['arguments'])
+      } else {
+        argumentsExpr = exprToString(expr['arguments'])
+      }
```
Comment (marrony) [inline src/Expr.js:295]: not necessary too

File: src/Expr.js [hunk 8]
```diff
@@ -169,35 +228,114 @@ var exprToString = function(expr, caller) {
...
+  }
+
+  if ('call' in expr) {
+    var argumentExpr = exprToString(expr['arguments'])
+
+    // Print when arguments are an object
+    if (
+      typeof expr['arguments'] === 'object' &&
+      !Array.isArray(expr['arguments']) &&
+      !expr['arguments']['raw']
```
Comment (marrony) [inline src/Expr.js:300]: that condition should never happen, you will never receive a vanilla object here, objects in our protocol are wrapped in `{"object": {...}}` function, and again, it will be handled in the beginning of the function

File: test/values.test.js [hunk 9]
```diff
@@ -501,12 +538,176 @@ describe('Values', () => {
...
+        call: {
+          function: 'double',
+        },
+      }),
+      `Query(Call(Function("double"), [1, 2, 3]))`
+    )
+
+    assertPrint(
+      new Query({
+        arguments: {
```
Comment (marrony) [inline test/values.test.js:701]: that argument is not correct, objects are wrapped in the protocol, the correct would be
```
new Query({
  arguments: {
    object: {
      one: 1,
      two: 2,
      three: 3
    }
  },
  call: {
    function: 'double'
  }
})
```

File: src/Expr.js [hunk 10]
```diff
@@ -169,35 +228,114 @@ var exprToString = function(expr, caller) {
...
+
+    return 'Call(' + exprToString(expr['call']) + ', ' + argumentExpr + ')'
+  }
+
   // Versioned queries/lambdas will have an api_version field.
-  // We want to prevent it from being parsed and displayed as:
-  // Query(ApiVersion("3", "X", Var("X")))
+  // We want to prevent it from being parsed and displayed.
   var keys = Object.keys(expr).filter(
     expression => expression !== 'api_version'
```
Comment (marrony) [inline src/Expr.js:311]: given lambdas are already handled before this point, that filter is not necessary

File: src/Expr.js [hunk 11]
```diff
@@ -169,35 +228,114 @@ var exprToString = function(expr, caller) {
...
     var v = expr[k]
     return exprToString(v, fn)
   })
 
-  var shouldReverseArgs = ['filter', 'map', 'foreach'].some(function(fn) {
-    return fn in expr
-  })
+  // Return string without args to avoid showing null args
+  if (args.length === 1 && args[0] === 'null') {
+    return fn + '()'
```
Comment (marrony) [inline src/Expr.js:335]: this will cause incorrect behaviors, not all functions with 1-arity that receives `null` should be rendered as `fn()`, for example, `IsNumber(null)` as is will be rendered as `IsNumber()`

File: src/Expr.js [hunk 12]
```diff
@@ -169,35 +228,114 @@ var exprToString = function(expr, caller) {
...
   )
+
   var fn = keys[0]
 
-  // For FQL functions with special formatting concerns, we
-  // use the specialCases object above to define their casing.
-  if (fn in specialCases) fn = specialCases[fn]
+  // Handle functions with an arity greater than 1
+  if (keys.length > 1) {
+    fn = getFunctionFromKeys(keys)
```
Comment (marrony) [inline src/Expr.js:318]: I don't think I understood the purpose of function `getFunctionFromKeys`, all the functions with arity > 1 will be handled before this point to avoid order problems

File: test/values.test.js (hunk 9 above)
Comment (fauna-ryanharris) [reply test/values.test.js:701]: Ah, okay. Thank you. I was missing the `object` 🤦

File: src/Expr.js (hunk 11 above)
Comment (fauna-ryanharris) [reply src/Expr.js:320]: Can you clarify this a bit more? My understanding was that whenever a function has one argument which is `null`, we don't want to render that.

File: src/Expr.js [hunk 13]
```diff
@@ -169,35 +218,92 @@ var exprToString = function(expr, caller) {
...
+    return (
+      'Call(' +
+      exprToString(expr['call']) +
+      ', ' +
+      exprToString(expr['arguments']) +
+      ')'
+    )
+  }
+
+  if ('databases' in expr) {
```
Comment (fauna-ryanharris) [inline src/Expr.js:293]: @marrony I added this to handle the case where tests were failing due to `Databases(null)` being rendered. I'm assuming we'd also want to include `Collections()`, `Documents()`, etc. as well, right?

Comment (marrony) [reply src/Expr.js:293]: yes, that's correct

File: src/Expr.js [hunk 14]
```diff
@@ -74,60 +74,109 @@ var specialCases = {
...
+}
 
-  if (type === 'undefined') {
-    return 'undefined'
+/**
+ *
+ * @param {*} expr A FQL expression
+ * @returns {Boolean} Indicates if arguments should be reversed
+ */
+function shouldReverseArgs(expr) {
```
Comment (marrony) [inline src/Expr.js:143]: also, once `exprToString` is finished, that function won't necessary, `filter` and `foreach` is already handled there, `map` should be too

File: src/Expr.js (hunk 11 above)
Comment (marrony) [reply src/Expr.js]: only 0-arity functions should drop null, like `Now()`. `IsNumber(x)` is 1-arity function and whatever argument was passed in should be preserved. You need to look at the wire to protocol to understand better, given `{}` represents a function call and the fields represents the function/arguments name, `Now()` function becomes `{"now": null}` in the wire, the null value is there because something must be in the value. On the other hand, `IsNumber(x)` is a 1-arity function, and it becomes `{"is_number": x}` in the wire

File: src/Expr.js (hunk 13 above)
Comment (marrony) [reply src/Expr.js:293]: bear in mind that `Databases()` and its siblings could receive an argument, the scope, so you should check if `expr[databases]` is not null and render it too like `Databases(Database('child-db'))`

File: test/values.test.js [hunk 15]
```diff
@@ -501,12 +538,178 @@ describe('Values', () => {
...
+        collection: [1, 2, 3],
+        api_version: '3',
+        filter: {
+          expr: {
+            equals: [0, { modulo: [{ var: 'i' }, 2] }],
+          },
+          lambda: 'i',
+        },
+      }),
+      'Query(Filter([1, 2, 3], Lambda("i", Equals(0, Modulo(Var("i"), 2)))))'
```
Comment (marrony) [inline test/values.test.js:597]: although this is kinda acceptable for testing purposes, you will never receive a `@query` type with a `Filter` expression. Queries types only wrap Lambdas

File: test/values.test.js [hunk 16]
```diff
@@ -501,12 +538,178 @@ describe('Values', () => {
...
+        lambda: '_',
+      }),
+      'Query(Lambda("_", Match(Index("idx"), 10)))'
+    )
+
+    // Filter an array
+    assertPrint(
+      new Query({
+        collection: [1, 2, 3],
+        api_version: '3',
```
Comment (marrony) [inline test/values.test.js:589]: Filter expressions doesn't have `api_version` argument.

File: src/Expr.js (hunk 14 above)
Comment (fauna-ryanharris) [reply src/Expr.js:143]: To make sure I understand, was this originally here to force the order of arguments for `filter`, `foreach` and `map` but now we don't need this function since our `if` statements are now handling constructing the queries properly?

File: test/values.test.js (hunk 15 above)
Comment (fauna-ryanharris) [reply test/values.test.js:597]: @marrony I noticed this when testing queries in the webshell that no `@query` was coming back for certain things, but my assertions were failing for some reason without `Query()`. I'm happy to go back and update these if you'd like 👍

File: test/values.test.js (hunk 16 above)
Comment (fauna-ryanharris) [reply test/values.test.js:589]: Good call. I'll remove this 👍

File: src/Expr.js (hunk 14 above)
Comment (marrony) [reply src/Expr.js:143]: That’s correct

File: test/values.test.js (hunk 15 above)
Comment (marrony) [reply test/values.test.js:596]: I don’t think it’s necessary to change, it was just a heads up to make sure you add/keep the lambdas tests

Comment (fauna-ryanharris) [reply test/values.test.js:596]: In the future, would the best way to do it be using `new Expr()`?

Comment (marrony) [reply test/values.test.js:596]: `Query` extends from `Expr`

File: test/values.test.js [hunk 17]
```diff
@@ -501,12 +538,228 @@ describe('Values', () => {
...
+        collection: { paginate: { documents: { collection: 'orders' } } },
+        foreach: {
+          lambda: 'doc',
+          expr: {
+            update: { var: 'doc' },
+            params: { object: { data: { object: { createdBy: 'Ryan' } } } },
+          },
+        },
+      }),
+      `Query(Foreach(Paginate(Documents(Collection("orders"))), Lambda("doc", Update(Var("doc"), {data: {createdBy: "Ryan"}}))))`
```
Comment (fauna-ryanharris) [inline test/values.test.js:664]: @marrony I hit some of your notes re: Databases, Collections and Documents, however, this test is now failing. For some reason, the Documents() never displays the Collection("orders") inside it.

I've tracked the issue to the `if ('paginate' in expr)` block of Expr.js, but am not sure why it is disregarding my current logic (which I think should work) 🤔

File: src/Expr.js (hunk 14 above)
Comment (fauna-ryanharris) [reply src/Expr.js:143]: Some tests are now failing if I remove this function call.

File: src/Expr.js [hunk 18]
```diff
@@ -169,35 +230,108 @@ var exprToString = function(expr, caller) {
...
+  if ('collections' in expr) {
+    if (!expr['database']) return 'Collections()'
+
+    return 'Collections(' + exprToString(expr['collections']) + ')'
+  }
+
+  if ('documents' in expr) {
+    console.log('in documents block')
+    console.log(expr)
+    if (!expr['collection']) return 'Documents()'
```
Comment (marrony) [inline src/Expr.js:320]: No need to test this, `collection` will never appear along with `documents` expression

File: src/Expr.js [hunk 19]
```diff
@@ -169,35 +230,108 @@ var exprToString = function(expr, caller) {
...
+      'Call(' +
+      exprToString(expr['call']) +
+      ', ' +
+      exprToString(expr['arguments']) +
+      ')'
+    )
+  }
+
+  if ('databases' in expr) {
+    if (!expr['database']) return 'Databases()'
```
Comment (marrony) [inline src/Expr.js:306]: `expr['databases']`

File: src/Expr.js [hunk 20]
```diff
@@ -169,35 +230,108 @@ var exprToString = function(expr, caller) {
...
+
+    return 'Databases(' + exprToString(expr['databases']) + ')'
+  }
+
+  if ('collections' in expr) {
+    if (!expr['database']) return 'Collections()'
```
Comment (marrony) [inline src/Expr.js:312]: `expr['collections']`

File: src/Expr.js (hunk 14 above)
Comment (marrony) [reply src/Expr.js:143]: you should add `map` to the if wagon

File: test/values.test.js (hunk 15 above)
Comment (marrony) [reply test/values.test.js:596]: but it could probably be a good idea to start using `Expr` like those tests https://github.com/fauna/faunadb-js/pull/310/files#diff-04f5179b95b9dabc1a5d3bfa30234fffR716

File: src/Expr.js [hunk 21]
```diff
@@ -169,35 +225,115 @@ var exprToString = function(expr, caller) {
...
+  }
+
+  if ('collections' in expr) {
+    if (!expr['collections']) return 'Collections()'
+
+    return 'Collections(' + exprToString(expr['collections']) + ')'
+  }
+
+  if ('documents' in expr) {
+    if (!expr['documents']) return 'Documents()'
```
Comment (marrony) [inline src/Expr.js:313]: `Documents` functions always receive an argument, there is not such overload with 0-arity

File: src/Expr.js [hunk 22]
```diff
@@ -169,35 +225,115 @@ var exprToString = function(expr, caller) {
...
     var v = expr[k]
     return exprToString(v, fn)
   })
 
-  var shouldReverseArgs = ['filter', 'map', 'foreach'].some(function(fn) {
-    return fn in expr
-  })
-
-  if (shouldReverseArgs) {
+  if (shouldReverseArgs(expr)) {
```
Comment (marrony) [inline src/Expr.js:336]: now that you handled all the edge case, that function call and the function itself is not necessary

File: test/values.test.js [hunk 23]
```diff
@@ -501,12 +538,228 @@ describe('Values', () => {
...
+        },
+      }),
+      'Query(Documents(Collection("lettuce")))'
+    )
+
+    assertPrint(
+      new Query({
+        documents: null,
+      }),
+      'Query(Documents())'
```
Comment (marrony) [inline test/values.test.js:762]: that assertion is not correct, if `documents` field is null, then it should render `Documents(null)`

Comment (fauna-ryanharris) [reply test/values.test.js:762]: @marrony Ah, okay. I was wondering why that was failing 😅

File: src/Expr.js (hunk 14 above)
Comment (fauna-ryanharris) [reply src/Expr.js:143]: Done 👍

File: src/Expr.js [hunk 24]
```diff
@@ -169,38 +214,113 @@ var exprToString = function(expr, caller) {
...
+      ')'
+    )
+  }
 
-  // For FQL functions with special formatting concerns, we
-  // use the specialCases object above to define their casing.
-  if (fn in specialCases) fn = specialCases[fn]
+  if ('containsstr' in expr) {
+    return (
+      'ContainsStr(' +
```
Comment (marrony) [inline src/Expr.js:229]: are you planning to add other string functions here? `containsstr` is already handled by `specialCases` map, if you plan to add those functions here, you can remove from there

Comment (fauna-ryanharris) [reply src/Expr.js:229]: @marrony I guess we could remove them from the special cases map and add all the `Contains` functions here since we need to make sure the argument order is constructed properly 👍 

What do you think?

**Comment by ecwyne** [comment]
@fauna-ryanharris I just wanted to double check on something.

Currently I can do the following:
```js
q.Let({
    one: q.Add(0, 1),
    two: q.Add(q.Var('one'), 1),
    three: q.Add(q.Var('two'), 1),
}, ...)
```

Will this still be possible with these changes?

**Comment by fauna-ryanharris** [comment]
> @fauna-ryanharris I just wanted to double check on something.
> 
> Currently I can do the following:
> 
> ```js
> q.Let({
>     one: q.Add(0, 1),
>     two: q.Add(q.Var('one'), 1),
>     three: q.Add(q.Var('two'), 1),
> }, ...)
> ```
> 
> Will this still be possible with these changes?

@ecwyne I believe so as I did not change any of the `let` functionality. @marrony Is there any reason Eric wouldn't be able to continue doing this?

**Comment by ecwyne** [comment]
To clarify, I only ask because the order of the properties on the `Let` object matters

**Comment by marrony** [comment]
@ecwyne this change has nothing to do with fql syntax per se, this change is only related to the fauna shell and webshell when the driver render a lambda type, it should produce an executable code so the user can copy and paste, or edit the lambda predicates on abac ui. @fauna-ryanharris Eric brought an interesting point, you need to make sure the let doesnt change the order of the variable declared.

**Comment by BrunoQuaresma** [comment]
@ryancharris I think add a test verifying the order on let params can be enough to ensure that.

File: test/query.test.js [hunk 25]
```diff
@@ -225,6 +225,17 @@ describe('query', () => {
           api_version: '3',
         })
       ),
+      assertQuery(
+        query.Let(
```
Comment (fauna-ryanharris) [inline test/query.test.js:229]: @marrony @BrunoQuaresma Changing the order of the fields in this object causes the test to fail. However, I don't think we've necessarily introduced this as a bug 🤔 Thoughts?

Comment (BrunoQuaresma) [reply test/query.test.js:229]: This test looks good :D

File: test/values.test.js [hunk 26]
```diff
@@ -501,12 +538,265 @@ describe('Values', () => {
...
+    assertPrint(
+      new Query({
+        let: {
+          three: { add: [{ var: 'two' }, 1] },
+          two: { add: [{ var: 'one' }, 1] },
+          one: { add: [0, 1] },
+        },
+        in: { var: 'three' },
+      }),
+      'Query(Let({three: Add(Var("two"), 1), two: Add(Var("one"), 1), one: Add(0, 1)}, Var("three")))'
```
Comment (marrony) [inline test/values.test.js:799]: I would add a test with arrays to make sure we are printing it correctly, something like: `Query(Let([{x: 10}, {y: 20}], Add(Var('x'), Var('y'))))`

Comment (fauna-ryanharris) [reply test/values.test.js:799]: @marrony This query doesn't work in the Webshell. From my reading of the FQL cheatsheet, it seems the second arg to `Let` needs to be a `Var` not an `Add`.

Comment (marrony) [reply test/values.test.js:799]: the second argument of `Let` is an expression, so it basically can be anything

Comment (fauna-ryanharris) [reply test/values.test.js:799]: @marrony Got it. I expected it to work, but I am getting this error:
```
Error: [
  {
    "position": [
      "query"
    ],
    "code": "invalid expression",
    "description": "No form/function found, or invalid argument keys: { let, in }."
  }
]
```

Comment (marrony) [reply test/values.test.js:799]: this worked for me
```javascript
Query(Lambda(x => Let([{y: x}], Add(1, Var("y")))))
```

**Comment by fauna-ryanharris** [comment]
@marrony Double checking to see if you have any other notes before I merge this in 👍

**Comment by marrony** [comment]
well, given we are now rendering lambdas so that users can persist and later edit it on the UI, ideally we should add tests with all functions in the API

**Comment by fauna-ryanharris** [comment]
> well, given we are now rendering lambdas so that users can persist and later edit it on the UI, ideally we should add tests with all functions in the API

cc: @danieltodonnell @nathaniel-may Should we include this in within the scope of this PR or spin the work off into its own ticket?

**Comment by danieltodonnell** [comment]
@ryancharris  My vote is to spin it into it's own card.  AS it is, THIS card was already overreach from the APIv3 work.  Let's let Product prioritize it.

**Comment by marrony** [comment]
sounds good to me, this PR let the current code on a par with the previous one, so any "bugs" as of now will likely exist anyway

**Comment by fauna-ryanharris** [comment]
Thanks for all your help with this PR, @marrony . Much appreciated 🙏 I will queue up another ticket for adding tests to cover the Lambdas 👍

**Comment by fauna-ryanharris** [comment]
> Thanks for all your help with this PR, @marrony . Much appreciated 🙏 I will queue up another ticket for adding tests to cover the Lambdas 👍

Here is the follow-up ticket: https://faunadb.atlassian.net/browse/FE-728
